pkg/job/job.go

```diff
@@ -0,0 +1,122 @@
+package job
+
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+)
+
+// job represents a process with owner and resource limits in any execution
+// state.
+type job struct {
+	mutex  sync.Mutex // protects concurrent access to status which contains mutable state
+	status Status
+
+	cmd    *exec.Cmd
+	owner  string
+	cgroup string
+}
+
+// newJob creates a new job with the given id, command, owner, limits and
+// cgroup.
+func newJob(owner, id string, command string, args []string, limits Limits, cgroup string) (*job, error) {
+	cmd, err := newStartedCmd(id, command, args, limits, cgroup)
+	if err != nil {
+		return nil, err
+	}
+	return &job{
+		status: Status{
+			ID:       id,
+			Command:  command,
+			Args:     args,
+			Started:  time.Now(),
+			Running:  true,
+			ExitCode: NotTerminated,
+		},
+		cmd:    cmd,
+		owner:  owner,
+		cgroup: cgroup,
+	}, nil
+}
+
+// isRunning synchronously returns the running status of the job.
+func (j *job) isRunning() bool {
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+	return j.status.Running
+}
+
+// getStatus synchronously creates a copy of the Status of the current to be
+// used to returned to the client.
+func (j *job) getStatus() Status {
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+	return j.status
+}
+
+// stop stops the job with a `SIGKILL` signal.
+func (j *job) stop() error {
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+	if !j.status.Running {
+		slog.Info("job already stopped", "id", j.status.ID)
+		return nil
+	}
+	// Write "1" to <job-cgroup>/cgroup.kill to kill job process and all its children
+	if err := writeCgroupFile(j.cgroup, "cgroup.kill", "1"); err != nil {
+		return fmt.Errorf("%w: cannot write to cgroup.kill %q: %q, %w", ErrJobStop, j.status.ID, j.cgroup, err)
+	}
+	return nil
+}
+
+// wait waits for the job to finish, updates the job status and deletes its
+// cgroups. It must only be called once per job.
+func (j *job) wait() {
+	waitErr := j.cmd.Wait()
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+	j.status.Running = false
+	j.status.Stopped = time.Now()
+	exitErr := &exec.ExitError{}
```

Comment:
nit: there's no need to actually allocate this error
```suggestion
	var exitErr *exec.ExitError
```

Reply:
Done.